kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala

```diff
@@ -0,0 +1,126 @@
+import org.apache.kyuubi.engine.ApplicationOperation.{APP_ERROR_KEY, APP_ID_KEY, APP_NAME_KEY, APP_STATE_KEY}
+import org.apache.kyuubi.engine.KubernetesApplicationOperation._
+
+class KubernetesApplicationOperation extends ApplicationOperation with Logging {
+
+  @volatile
+  private var kubernetesClient: KubernetesClient = _
+
+  override def initialize(conf: KyuubiConf): Unit = {
+    info("Start Initialize Kubernetes Client.")
+    val contextOpt = conf.get(KUBERNETES_CONTEXT)
+    if (contextOpt.isEmpty) {
+      warn("Skip Initialize Kubernetes Client, because of Context not set.")
+      return
+    }
+    try {
+      kubernetesClient = new DefaultKubernetesClient(Config.autoConfigure(contextOpt.get))
+      info(s"Initialized Kubernetes Client connect to: ${kubernetesClient.getMasterUrl}")
+    } catch {
+      case e: KubernetesClientException =>
+        error("Fail to init KubernetesClient for KubernetesApplicationOperation", e)
+    }
+  }
+
+  override def isSupported(clusterManager: Option[String]): Boolean = {
+    kubernetesClient != null && clusterManager.nonEmpty &&
+    clusterManager.get.toLowerCase.startsWith("k8s")
+  }
+
+  override def killApplicationByTag(tag: String): KillResponse = {
+    if (kubernetesClient != null) {
+      debug(s"Deleting application info from Kubernetes cluster by $tag tag")
+      try {
+        // Need driver only
+        val operation = findDriverPodByTag(tag)
+        val pod = operation.list().getItems.get(0)
+        (
+          operation.delete(),
+          s"Operation of deleted appId: " +
+            s"${pod.getMetadata.getName} is completed")
+      } catch {
+        case e: Exception =>
+          (false, s"Failed to terminate application with $tag, due to ${e.getMessage}")
+      }
+    } else {
+      throw new IllegalStateException("Methods initialize and isSupported must be called ahead")
+    }
+  }
+
+  override def getApplicationInfoByTag(tag: String): Map[String, String] = {
+    if (kubernetesClient != null) {
+      debug(s"Getting application info from Kubernetes cluster by $tag tag")
+      try {
+        val operation = findDriverPodByTag(tag)
+        val pod = operation.list().getItems.get(0)
+        val res = Map(
+          // Can't get appId, get Pod UID instead.
+          APP_ID_KEY -> pod.getMetadata.getUid,
+          APP_NAME_KEY -> pod.getMetadata.getName,
+          APP_STATE_KEY -> pod.getStatus.getPhase,
+          APP_ERROR_KEY -> pod.getStatus.getReason)
+        debug(s"Successfully got application info by $tag: " + res.mkString(", "))
+        res
+      } catch {
+        case e: Exception =>
+          error(s"Failed to get application with $tag, due to ${e.getMessage}")
+          null
+      }
+    } else {
+      throw new IllegalStateException("Methods initialize and isSupported must be called ahead")
+    }
+  }
+
+  private def findDriverPodByTag(tag: String): FilterWatchListDeletable[Pod, PodList] = {
```

Comment:
what if the deploy-mode is client ?

Reply:
Currently, Cluster mode is only available for On Kubernetes. We do not consider Client mode to be true On Kubernetes. It is not delivered to Resource Manager. I can add a layer of restriction at `isSupported`.